File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> hmap;
        hmap['{'] = '}';
        hmap['['] = ']';
        hmap['('] = ')';

        stack<char> container;

        for(char c : s) {
            if (hmap.contains(c)) {
                container.push(c);
            } else {
                if (container.size() == 0)
                    return false;
                char top = container.top();
                container.pop();
                if (c != hmap[top])
                    return false;
            }
        }

        return container.size() > 0 ? false : true;
    }
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        checker = { ')' : '(', '}' : '{', ']' : '[' }

        for c in s:
            if c == '(' or c == '{' or c == "[":
                stack.append(c)
            else:
                if stack:
                    top = stack.pop()
                    if top != checker[c]:
                        return False
                else:
                    stack.append(c)
        
        if stack:
            return False
            
        return True

